pkg/proxy/proxy.go

```diff
@@ -391,18 +392,52 @@ func (p *Proxy) addUserContext() echo.MiddlewareFunc {
 				return next(ctx)
 			}
 
-			userID, username, err := p.extractUserID(ctx.Request())
+			token, err := p.extractUserToken(ctx.Request())
 			if err != nil {
 				return crterrors.NewUnauthorizedError("invalid bearer token", err.Error())
 			}
-			ctx.Set(context.SubKey, userID)
-			ctx.Set(context.UsernameKey, username)
+			ctx.Set(context.SubKey, token.Subject)
+			ctx.Set(context.UsernameKey, token.PreferredUsername)
+			ctx.Set(context.EmailKey, token.Email)
 
 			return next(ctx)
 		}
 	}
 }
 
+// ensureUserIsNotBanned rejects the request if the user is banned
+func (p *Proxy) ensureUserIsNotBanned() echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(ctx echo.Context) error {
+			errorResponse := func(err *crterrors.Error) error {
+				ctx.Logger().Error(errs.Wrap(err, "workspace list error"))
+				ctx.Response().Writer.Header().Set("Content-Type", "application/json")
+				ctx.Response().Writer.WriteHeader(int(err.Code))
+				return json.NewEncoder(ctx.Response().Writer).Encode(err.Status)
+			}
+
+			email := ctx.Get(context.EmailKey).(string)
+			if email == "" {
+				return errorResponse(crterrors.NewUnauthorizedError("unauthenticated request", "anonymous access is not allowed"))
+			}
+
+			// retrieve banned users
+			uu, err := p.app.InformerService().BannedUsersByEmail(email)
+			if err != nil {
+				return errorResponse(crterrors.NewInternalError(errs.New("unable to retrieve user"), "could not define ban status"))
+			}
+
+			// if a matching Banned user is found, then user is banned
+			if len(uu) > 0 {
+				return errorResponse(crterrors.NewForbiddenError("user is banned", "user is banned"))
```

Comment:
We usually don't indicate that user is banned and return a generic forbidden error message for banned users. Just want to give as little details to banned users as possible. Not sure if it's right approach or not though. WDYT?
But it's not a big deal anyway.

Reply:
I think it's good for now to hide details. The worst think that can happen is that banned users cannot tell they are banned so they could start opening tickets complaining about the service not working properly.

I don't see this becoming a problem in the short term and given that this is the approach used up to now, I'd say to hide details also in this PR. If we start to have tickets for this behavior or the team decides to change the approach, we can create a new story.

fixed in 6c12ec8d0b69a7c6424897acb0be6a8193f9da6a